File: asm/src/asm.cpp
#include "asm.h"
#include <stdio.h>

string Error = "Error";
string Warning = "Warning";
string InvalidDeclaration = ":InvalidDeclaration";
string Unknown = ":Unknown";
string PreprocessStatement = ":PreprocessStatement";
string Constant = ":Constant";
string BadCommandLine = ":BadCommandLine";
string InvalidOption = ":InvalidOption";
string NotPresent = ":NotPresent";


string self;

void RaiseError(string e, string m) {
    printf("%s: %s: %s\n", self, Error + e, m);
}

int main(int argc, char ** argv) {
    // convert argv to vector<string>
    vector<string> args;
    for (int i=1; i<argc; i++) args.append(argv[i]);

    // get executable name
    self = argv[0];
    vector<string> path = self.split('\\');
    self = path[path.count-1];

    string infile = "";
    string outfile = "";
    string text;
    int o = 1;
    bool error = false;
    Assembler assembler;
    vector<byte> bin;

    // general args
    for (int i=0; i<args.count; i++) {
        string elem = args[i];
        if (elem.startswith('-')) {
            if (elem == "-O1") o = 1; // normal
            else if (elem == "-O2") o = 2; // compact operand sizes
            else {
                if (elem != "-o") {
                    char * buf;
                    sprintf(buf, "Unknown command line option '%s'", elem);
                    RaiseError(BadCommandLine + InvalidOption, string(buf));
                    error = true;
                }
            }
        }
    }

    // multi param args
    for (int i=0; i<args.count; i++) {
        if (args[i] == "-o") {
            outfile = args[++i];
        } else if (!args[i].startswith('-')) {
            infile = args[i];
        }
    }

    if (infile == "") {
        RaiseError(BadCommandLine + NotPresent, "no input files");
        error = 1;
    }

    // infile
    if (outfile == "") {
        outfile = infile.split('.')[0] + ".bin";
    }

    // arg-parse errors
    if (error) return 1;

    // read infile
    text = file::ReadAllText(infile);
    printf("%s\n", text);
    return 0;

    // assemble
    bin = assembler.Assemble(text);
    if (bin.count == 0) return 1;

    // write to outfile
    file::WriteAllBytes(outfile, bin);

    return 0;
}

File: ccc/src/asm.py
import sys

this: str = ""

class Error:...

class Warn:...

class InvalidDeclaration(Error):...

class Unknown(Error):...

class PreprocessStatement(Unknown):...

class DefineConstant(InvalidDeclaration):...

class BadCommandLine(Error):...

class InvalidOption(BadCommandLine):...

def all_member(cl: (Error | Warn)) -> list[str]:
    classes = []
    strClasses = []

    for base in cl.__bases__:
        if base == object:
            break;

        classes.extend(all_member(base))

    classes.append(cl)
    return classes

def RaiseError(error: Error, message: str) -> None:
    print("%s: %s: %s" % (this, ":".join([i.__name__ for i in all_member(error.__class__)]), message))

def RaiseWarning(warning: Warn, message: str) -> None:
    print("%s: %s: %s" % (this, ":".join([i.__name__ for i in all_member(warning.__class__)]), message))

def IsNullOrWhitespace(string=None) -> bool:
    return not string or string.isspace()

class Assembler:
    def __init__(self) -> None:
        self.labels: dict[str, int] = {}
        self.constants: dict[str, int] = {}
        self.pointers: dict[str, int] = {}
        self.errors: bool = False

    def PreParse(self, text: str) -> str:
        section = "glob"
        parsedText = ""

        for line in text.split('\n'):
            if line.startswith('.'):
                match line.split()[0]:
                    case ".const":
                        if (section == "glob"):
                            tokens = line.split(' ')
                            self.constants[tokens[1]] = int(tokens[2])
                        else:
                            RaiseError(DefineConstant, "Defining constant outside global scope")
                        break
                    case ".data":
                        section = "data"
                        break
                    case ".text":
                        section = "text"
                        break
                    case _:
                        RaiseError(PreprocessStatement(), "Unknown preprocess statement '%s'; skipping" % (line.split()[0]))
                        break

    def StripWhitespace(self, lines: list[str]) -> list[str]:
        strippedText = []

        for line in lines:
            line = line.split(';')[0]
            line = " ".join(line.split())
            if not IsNullOrWhitespace(line):
                strippedText.append(line)

        print("\n".join(strippedText))

        return strippedText        

    def ParseDisassemble(self) -> bytearray:
        binary = bytearray()

        return binary

    def Assemble(self, text: str):
        binary = bytearray()
        text = text.split('\n')

        text = "\n".join(self.StripWhitespace(text))
        text = self.PreParse(text)

        if self.errors:
            return None

        binary = self.ParseDisassemble()

        return binary

def main(argv: list[str]):
    argv = argv[1:]
    global this
    infile: str = ""
    outfile: str = None
    text: str = ""
    opt: int = 1
    error: bool = False
    assembler: Assembler = Assembler()
    binary: bytearray = bytearray()

    this = argv[0].split('/')[-1]

    for i in argv:
        if i.startswith('-'):
            match i:
                case "-O1":
                    opt = 1
                    break
                case "-O2":
                    opt = 2
                case _:
                    if i != "-o":
                        RaiseError(InvalidOption(), "Unknown command line option '%s'" % (i))
                        error = True

    i = 0
    while (i < (len(argv)-1)):
        if argv[i] == "-o":
            outfile = argv[i+1]
            i+=1
        elif not argv[i].startswith('-'):
            infile = argv[i]
        i+=1

    if outfile is None:
        outfile = infile.split('.')[0] + ".bin"

    if error:
        return

    with open(infile, "r") as f:
        text = f.read()

    print(text)
    binary = assembler.Assemble(text)
    if binary is None:
        return

    with open(outfile, "wb") as f:
        f.write(binary)

if __name__ == "__main__":
    main(sys.argv)
